### PR TITLE
Fix concurrency callbacks

### PR DESCRIPTION
**Pull request by boaz-codota**

### CODE REVIEW

File: src/binary/InnerBinary.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  public init(proc: child_process.ChildProcess, readline: ReadLine): void {
+    this.proc = proc;
+    this.onceReader = new OnceReader(readline);
+  }
+
+  public async request<T, R = unknown>(
+    request: R,
+    timeout = 1000
+  ): Promise<T | null | undefined> {
+    try {
```
Comment (yonip23) [inline src/binary/InnerBinary.ts:24]: remove `try catch` clause as `Binary.ts` already `try-catch`ing it

File: src/binary/InnerBinary.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    timeout: number
+  ): Promise<UnkownWithToString> {
+    return new Promise<UnkownWithToString>((resolve, reject) => {
+      setTimeout(() => {
+        reject(new Error("Binary request timed out."));
+      }, timeout);
+
+      this.onceReader?.read(resolve);
+
+      this.proc?.stdin.write(
```
Comment (yonip23) [inline src/binary/InnerBinary.ts:45]: dont we want to fail upon missing `onceReader` or `proc`? would be better to use `!` in my opinion

File: src/binary/mockedRunProcess.ts [hunk 3]
```diff
@@ -1,63 +1,79 @@
...
+
+      return !!completionRequest?.request?.Features;
+    },
+    result: {
+      enabled_features: [Capability.ALPHA_CAPABILITY],
+    },
+  });
+}
+
+function mockBinaryRequest(): void {
```
Comment (yonip23) [inline src/binary/mockedRunProcess.ts:54]: rename to `mockBinaryRequestListener`?

File: src/test/suite/notification.test.ts [hunk 4]
```diff
@@ -101,38 +114,34 @@ suite("Should poll notifications", () => {
...
-        x.write(
-          `{\"version\":\"${API_VERSION}\",\"request\":{\"NotificationAction\":{\"id\":\"ANOTHER_NOTIFICATION_ID\",\"message\":\"ANOTHER_MESSAGE\",\"notification_type\":\"promo\",\"state\":null}}}\n`,
-          "utf8"
-        ),
-      TypeMoq.Times.once()
-    );
-    showInformationMessage.restore();
+    verify(
+      stdinMock.write(
+        `{"version":"${API_VERSION}","request":{"NotificationAction":{"id":"DIFFERENT_NOTIFICATION_ID","selected":"AN_OPTION_KEY","message":"A_MESSAGE","notification_type":"promo","actions":["None"],"state":null}}}\n`,
```
Comment (yonip23) [inline src/test/suite/notification.test.ts:134]: extract expected string to const / function

File: src/test/suite/notification.test.ts [hunk 5]
```diff
@@ -101,38 +114,34 @@ suite("Should poll notifications", () => {
...
+        "utf8"
+      )
+    ).once();
+    // );
```
Comment (yonip23) [inline src/test/suite/notification.test.ts:138]: remove

File: src/test/suite/notification.test.ts [hunk 6]
```diff
@@ -101,38 +114,34 @@ suite("Should poll notifications", () => {
...
+    verify(
+      stdinMock.write(
```
Comment (yonip23) [inline src/test/suite/notification.test.ts:140]: extract expected string to const / function

File: src/test/suite/utils/completion.utils.ts [hunk 7]
```diff
@@ -15,42 +10,42 @@ export type AutocompleteRequest = BinaryGenericRequest<{
...
+//     .setup((x) => x.on("line", TypeMoq.It.isAny()))
+//     .callback((_x, callback: (line: string) => void) => {
+//       if (!requestHappened) {
+//         callback("null");
+//       } else {
+//         callback(JSON.stringify(response[requestAnswered] || null));
+//         requestAnswered += 1;
+//       }
+//     });
+// }
```
Comment (yonip23) [inline src/test/suite/utils/completion.utils.ts:48]: # EXCUSE ME

File: src/binary/mockedRunProcess.ts [hunk 8]
```diff
@@ -1,63 +1,79 @@
...
+      matchingItem ? response(request, matchingItem.result) : "null"
+    );
+
+    return true;
+  });
 }
 
+function response(request: string, result: Result) {
+  return JSON.stringify(
+    isFunction(result) ? (result as ResultFunction)(request) : result
```
Comment (yonip23) [inline src/binary/mockedRunProcess.ts:77]: unless i missed something, i don't see any usage of `Result` as a function, and it only seem to complicate the usage here. Just  use a simple string, and this whole function disappears.

**Review by yonip23 — APPROVED** [review]
This is quite a lot of changes, hope i didn't miss anything here. (i know most of them are linting, but still...)

Anyway, had some comments, besides that lgtm

File: src/binary/InnerBinary.ts (hunk 2 above)
Comment (boaz-codota) [reply src/binary/InnerBinary.ts:45]: Either way, it won't crash as you expect, so I don't see the point in using `!` over `?`.

File: src/binary/InnerBinary.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
+import * as child_process from "child_process";
+import { ReadLine } from "readline";
+import { API_VERSION } from "../consts";
+import OnceReader from "./OnceReader";
+
+type UnkownWithToString = {
+  toString(): string;
+};
+
+export default class InnerBinary {
```
Comment (dimacodota) [inline src/binary/InnerBinary.ts:10]: maybe rename it to binaryRequestReader?

File: src/test/suite/utils/completion.utils.ts (hunk 7 above)
Comment (boaz-codota) [reply src/test/suite/utils/completion.utils.ts:48]: oops, removed

File: src/binary/mockedRunProcess.ts (hunk 8 above)
Comment (boaz-codota) [reply src/binary/mockedRunProcess.ts:77]: I used it as a function... See `setNotificationsResult` in `notification.utils.ts`

Comment (boaz-codota) [reply src/binary/mockedRunProcess.ts:77]: Such wow, much useful

File: src/binary/OnceReader.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+        throw new Error(
+          "Read a response from the engine before a request was written."
+        );
+      }
+
+      oldestCallback(line);
+    });
+  }
+
+  read(callback: (line: string) => void): void {
```
Comment (dimacodota) [inline src/binary/OnceReader.ts:22]: maybe call it once/onRead/onData ...

File: src/test/suite/utils/completion.utils.ts [hunk 11]
```diff
@@ -15,42 +10,42 @@ export type AutocompleteRequest = BinaryGenericRequest<{
...
-    .callback((x, callback: (line: string) => void) => {
-      if (!requestHappened) {
-        callback("null");
-      } else {
-        callback(JSON.stringify(response[requestAnswered] || null));
-        requestAnswered += 1;
-      }
-    });
-}
+// export function setCompletionResult(
```
Comment (dimacodota) [inline src/test/suite/utils/completion.utils.ts]: remove

**Review by dimacodota — APPROVED** [review]
lgtm

**Pull request merged**